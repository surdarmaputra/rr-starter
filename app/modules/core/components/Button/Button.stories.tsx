import type { Meta, StoryObj } from '@storybook/react';

import { Button, type ButtonProps } from './Button';

type Story = StoryObj<typeof Button>;

const meta: Meta = {
  component: Button,
  title: 'Core/Components/Button',
};

export default meta;

export const Default: Story = {
  args: {
    children: 'Button',
  },
  render: (args: ButtonProps) => <Button {...args} />,
};
