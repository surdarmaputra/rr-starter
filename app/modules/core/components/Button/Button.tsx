import { Slot } from '@radix-ui/react-slot';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '~/modules/core/libs/shadcn/utils';

import type { TestableComponentProps } from '../../types';

const buttonVariants = cva(
  `
    inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-semibold
    ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2
    focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none
    disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0
    dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300
  `,
  {
    variants: {
      variant: {
        primary: 'bg-primary text-white hover:bg-primary-hover',
        'primary-outlined':
          'border border-primary text-primary hover:bg-primary-hover hover:text-white',
        'primary-ghost': 'text-primary hover:bg-primary-hover hover:text-white',
        secondary: 'bg-secondary text-white hover:bg-secondary-hover',
        'secondary-outlined':
          'border border-secondary text-secondary hover:bg-secondary-hover hover:text-white',
        'secondary-ghost':
          'text-secondary hover:bg-secondary-hover hover:text-white',
        success: 'bg-success text-white hover:bg-success-hover',
        'success-outlined':
          'border border-success text-success hover:bg-success-hover hover:text-white',
        'success-ghost': 'text-success hover:bg-success-hover hover:text-white',
        warning: 'bg-warning text-white hover:bg-warning-hover',
        'warning-outlined':
          'border border-warning text-warning hover:bg-warning-hover hover:text-white',
        'warning-ghost': 'text-warning hover:bg-warning-hover hover:text-white',
        danger: 'bg-danger text-white hover:bg-danger-hover',
        'danger-outlined':
          'border border-danger text-danger hover:bg-danger-hover hover:text-white',
        'danger-ghost': 'text-danger hover:bg-danger-hover hover:text-white',
        link: 'text-textTitle underline-offset-4 hover:underline',
      },
      size: {
        default: 'h-10 px-4 py-2',
        sm: 'h-9 px-3',
        lg: 'h-11 px-8',
        icon: 'h-10 w-10',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'default',
    },
  },
);

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants>,
    TestableComponentProps {
  asChild?: boolean;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, size, asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : 'button';
    return (
      <Comp
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        {...props}
      />
    );
  },
);

Button.displayName = 'Button';

export { Button, buttonVariants };
