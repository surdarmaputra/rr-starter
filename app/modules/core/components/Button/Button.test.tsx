import { render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import { Button } from './Button';

describe('Button', () => {
  test('renders with data-testid', async () => {
    const dataTestId = 'sample-test-id';
    render(<Button data-testid={dataTestId} />);
    await screen.findByTestId(dataTestId);
  });

  test('renders snapshot', async () => {
    const { container } = render(<Button />);
    expect(container).toMatchSnapshot();
  });
});
