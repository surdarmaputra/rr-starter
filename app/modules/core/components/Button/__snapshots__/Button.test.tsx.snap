// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders snapshot 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-primary text-black hover:bg-primaryHover h-10 px-4 py-2"
  />
</div>
`;
